apiserver/keymanager/keymanager.go

```diff
@@ -44,58 +45,83 @@ type KeyManagerAPI struct {
 	state      *state.State
 	resources  facade.Resources
 	authorizer facade.Authorizer
-	canRead    func(string) bool
-	canWrite   func(string) bool
+	apiUser    *names.UserTag
 	check      *common.BlockChecker
 }
 
 var _ KeyManager = (*KeyManagerAPI)(nil)
 
 // NewKeyManagerAPI creates a new server-side keyupdater API end point.
 func NewKeyManagerAPI(st *state.State, resources facade.Resources, authorizer facade.Authorizer) (*KeyManagerAPI, error) {
-	// Only clients and environment managers can access the key manager service.
-	if !authorizer.AuthClient() && !authorizer.AuthModelManager() {
+	// Only clients and controllers can access the key manager service.
+	if !authorizer.AuthClient() && !authorizer.AuthController() {
 		return nil, common.ErrPerm
 	}
-	env, err := st.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	// For gccgo interface comparisons, we need a Tag.
-	owner := names.Tag(env.Owner())
-	// TODO(wallyworld) - replace stub with real canRead function
-	// For now, only admins can read authorised ssh keys.
-	canRead := func(user string) bool {
-		// Are we a machine agent operating as the system identity?
-		if user == config.JujuSystemKey {
-			_, ismachinetag := authorizer.GetAuthTag().(names.MachineTag)
-			return ismachinetag
-		}
-		return authorizer.GetAuthTag() == owner
-	}
-	// TODO(wallyworld) - replace stub with real canWrite function
-	// For now, only admins can write authorised ssh keys for users.
-	// Machine agents can write the juju-system-key.
-	canWrite := func(user string) bool {
-		// Are we a machine agent writing the Juju system key.
-		if user == config.JujuSystemKey {
-			_, ismachinetag := authorizer.GetAuthTag().(names.MachineTag)
-			return ismachinetag
-		}
-		// No point looking to see if the user exists as we are not
-		// yet storing keys on the user.
-		return authorizer.GetAuthTag() == owner
+	var apiUser *names.UserTag
+	if authorizer.AuthClient() {
+		u, _ := authorizer.GetAuthTag().(names.UserTag)
+		apiUser = &u
 	}
 	return &KeyManagerAPI{
 		state:      st,
 		resources:  resources,
 		authorizer: authorizer,
-		canRead:    canRead,
-		canWrite:   canWrite,
+		apiUser:    apiUser,
 		check:      common.NewBlockChecker(st),
 	}, nil
 }
 
+func (api *KeyManagerAPI) checkCanRead(sshUser string) error {
+	if err := api.checkCanWrite(sshUser); err == nil {
+		return nil
+	} else if err != common.ErrPerm {
+		return errors.Trace(err)
+	}
+	if api.apiUser == nil {
```

Comment:
`if api.apiUser == nil || sshUser == config.JujuSystemKey`

and please add a test that shows that a user cannot read the system key